Src/LogData.c

```diff
@@ -36,316 +73,209 @@ void writeToEEPROM(uint8_t* buffer, uint16_t bufferSize, uint16_t memAddress, ui
+ */
+void logEntryOnceRoutine(AllData* data)
 {
-    // Get card mounted
-    uint8_t mounted = 0;
-
-    while (!mounted)
-    {
-        // Keep trying, really need to log during this time
-        mounted = (f_mount(&fatfs, "SD:", 1) == FR_OK);
-
-        FlightPhase flightPhase = getCurrentFlightPhase();
-
-        if ( flightPhase != BURN &&
-                flightPhase != COAST &&
-                flightPhase != DROGUE_DESCENT)
-        {
-            // couldn't mount during important phases, too bad :(
-            return;
-        }
-    }
-
-    // Card mounted, start writing at high frequency
-    HAL_GPIO_WritePin(LED1_GPIO_Port, LED1_Pin, 1);
-    uint32_t prevWakeTime = osKernelSysTick();
-
-    for (;;)
-    {   // Grab
-        osDelayUntil(&prevWakeTime, FAST_LOG_DATA_PERIOD);
-
-        FlightPhase flightPhase = getCurrentFlightPhase();
-
-        if ( flightPhase != BURN &&
-                flightPhase != COAST &&
-                flightPhase != DROGUE_DESCENT)
-        {
-            // done important phases, unmount card and exit high frequency logging
-            break;
-        }
-
-        buildLogEntry(data, buffer);
-
-        if (f_open(&file, fileName, FA_OPEN_APPEND | FA_READ | FA_WRITE) == FR_OK)
-        {
-            f_puts(buffer, &file);
-            f_close(&file); // close to save the file
-        }
-    }
-
-    HAL_GPIO_WritePin(LED1_GPIO_Port, LED1_Pin, 0);
-    f_mount(NULL, "SD:", 1);
+    updateLogEntry(data);
+    writeLogEntryToEEPROM(currMemAddress);
 }
 
 void logDataTask(void const* arg)
 {
     AllData* data = (AllData*) arg;
-    char buffer[500];
-
-    sprintf(
-        buffer,
-        "accelX,"
-        "accelY,"
-        "accelZ,"
-        "gyroX,"
-        "gyroY,"
-        "gyroZ,"
-        "magnetoX,"
-        "magnetoY,"
-        "magnetoZ,"
-        "pressure,"
-        "temperature(100C),"
-        "combustionChamberPressure(1000psi),"
-        "oxidizerTankPressure(1000psi),"
-        "GPS_time,"
-        "GPS_latitude_degrees,"
-        "GPS_latitude_minutes,"
-        "GPS_longitude_degrees,"
-        "GPS_longitude_minutes,"
-        "GPS_altitude,"
-        "currentFlightPhase,"
-        "elapsedTime(ms),"
-        "softwareVersion\n"
-    );
-
+    initializeLogEntry();
+    uint32_t prevWakeTime;
+    clock_t beforeLogTime, afterLogTime, totalTime;
     for (;;)
     {
+        beforeLogTime = clock();
+        logEntryOnceRoutine();
+        afterLogTime = clock();
+        totalTime = (double)(afterLogTime - beforeLogTime) / CLOCKS_PER_SEC / 1000; //this also takes time!
+        prevWakeTime = osKernelSysTick();//assume it is atomic: runs in one cycle
         switch (getCurrentFlightPhase())
         {
             case BURN:
             case COAST:
             case DROGUE_DESCENT:
-                highFrequencyLogToSdRoutine(data, buffer);
+                osDelayUntil(&prevWakeTime, (uint32_t)(FAST_LOG_DATA_PERIOD - totalTime));
```

Comment:
Needs check for `totalTime > log period`. Otherwise negative time will be cast to near max uint32 size (500+ million years). Suggested solution: `(uint32_t)(max(LOG_PERIOD - totalTime, 0))`.

Also, to be more accurate, we should call the second time-get-function in the delay: `(uint32_t)(max(LOG_PERIOD - (osKernelSysTick() - beforeLogTime), 0))`. Don't need to worry about the clock rolling over, since that needs the clock to be running for 500+ million years.

Reply:
Don't know if `max(...)` is currently implemented. Otherwise we can use the following:
```
 #define max(a,b) \
   ({ __typeof__ (a) _a = (a); \
       __typeof__ (b) _b = (b); \
     _a > _b ? _a : _b; })
 ```
https://stackoverflow.com/questions/3437404/min-and-max-in-c

Reply:
Or find a better way 🤷